crates/uv-auth/src/providers.rs

```diff
@@ -47,3 +50,48 @@ impl HuggingFaceProvider {
         None
     }
 }
+
+/// The [`Url`] for the S3 endpoint, if set.
+static S3_ENDPOINT_URL: LazyLock<Option<Url>> = LazyLock::new(|| {
+    let s3_endpoint_url = std::env::var(EnvVars::UV_S3_ENDPOINT).ok()?;
+    let url = Url::parse(&s3_endpoint_url).expect("Failed to parse S3 endpoint URL");
+    Some(url)
+});
+
+/// A provider for authentication credentials for S3 endpoints.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct S3EndpointProvider;
+
+impl S3EndpointProvider {
+    /// Returns the credentials for the S3 endpoint, if available.
+    pub(crate) fn credentials_for(url: &Url) -> Option<DefaultSigner> {
+        if let Some(s3_endpoint_url) = S3_ENDPOINT_URL.as_ref() {
+            // Treat any URL on the same domain or subdomain (plus scheme and port) of the endpoint
+            // URL as available for signing.
+            if url.scheme() == s3_endpoint_url.scheme()
+                && url.port() == s3_endpoint_url.port()
+                && url.domain().is_some_and(|subdomain| {
+                    s3_endpoint_url.domain().is_some_and(|domain| {
+                        subdomain == domain
+                            || subdomain
+                                .strip_suffix(domain)
+                                .is_some_and(|prefix| prefix.ends_with('.'))
+                    })
+                })
+            {
```

Comment:
Do we just need to add `.without_subdomain` to `Realm`?

Reply:
It turns out that it's non-trivial to strip a subdomain, because you need the list eTLDs in order to figure out where the subdomain starts. For example, the URL could end in `.co.uk`, so it's not just "keep the last two segments".

Reply:
I added a `RealmRef::is_subdomain_of` method which does this. (Asking if it's a subdomain of something else is easier than stripping the subdomain, I think.)

Reply:
Oh interesting. That seems reasonable.